### PR TITLE
catalog: add support for replicating catalogs for reader virtual clusters

### LINKED ISSUE #129439: sql/catalog: extract and insert catalogs across tenants to enable PCR select from standby
Add support for extracting and replicating catalogs across tenants for PCR work. As a part of this work we will:

1. Extract a catologue out of offline tenant key span
2. Inject a catalogue into reader tenant key space
3. Move the catalogue timestamp in reader tenant key space.

See: https://cockroachlabs.atlassian.net/browse/CRDB-37521

Jira issue: CRDB-41544

Epic CRDB-37521

### PR DESCRIPTION
**Pull request by fqazi**
This patch adds logic for extracting and creating a PCR reader catalog. This patch will do the following:

1. Introduce a new catalog function catalog.SetupOrAdvanceStandbyReaderCatalog which can be used to advanced or setup a reader catalog from a source tenant.
2. Add logic to allow read only sequence access, so that SQL behaves correctly

Fixes: #129439

Note: The first commit in this PR can be ignored, and will be merged via a separate PR https://github.com/cockroachdb/cockroach/pull/130183

### CODE REVIEW

File: pkg/sql/catalog/replication/reader_catalog.go
Comment (msbutler) [inline pkg/sql/catalog/replication/reader_catalog.go:188]: perhaps we default error on an unrecognized type?

File: pkg/sql/catalog/replication/reader_catalog.go
Comment (msbutler) [inline pkg/sql/catalog/replication/reader_catalog.go:83]: nit: this comment is slightly misleading and i think the check could be stronger. Could the check just be:
```
existingDesc.GetParentID()  == desc.GetParentID()
```
i.e. we're asserting that all extracted user tables are existing user tables, and all extracted system table are a existing system tables.

Comment (msbutler) [reply pkg/sql/catalog/replication/reader_catalog.go:83]: we should probably also add an assertion that every extracted system table exists in the reader tenant.

File: pkg/sql/catalog/replication/reader_catalog_test.go
Comment (msbutler) [inline pkg/sql/catalog/replication/reader_catalog_test.go:91]: could you add a few test cases, where you rerun SetupOrAdvanceStandbyReaderCatalog with a new AOST after:
-  insert into t1
- create a new table and insert data into it
- conduct some schema change on it (e.g. add column in it)
- drop that table new table

**Review by msbutler — COMMENTED** [review]
this looks great! probably want a foundations reviewer on this as well.

File: pkg/sql/catalog/replication/reader_catalog.go
Comment (msbutler) [reply pkg/sql/catalog/replication/reader_catalog.go:83]: ^ actually, my check above :
```
existingDesc.GetParentID()  == desc.GetParentID()
```
could lead to false positives if the system database id is not hard coded at start up

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130184)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @azhu-crl, @msbutler, and @stevendanna)

___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747054480:-O6RBGWI-YO5FOVfIZXY:b2glbtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L35)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

i think this docstring could be clearer. The function scrapes the descriptors AOST from the passed in fromID, and updates the descriptors of the calling tenant. 
</blockquote></details>

Good point, updated this comment to:
// SetupOrAdvanceStandbyReaderCatalog when invoked inside the reader
// tenant will replicate the descriptors from the tenant specified
// by fromID. The replicated descriptors will be setup such that they
// will access data from fromID. If the descriptors are already replicated
// then this function will advance the timestamp.
___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747058805:-O6RBzZv7omvPMe3rE46:b-ed3vhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L56)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

stray comment?
</blockquote></details>

More of an incomplete thought:

			// Track which descriptors / namespaces that have been updated, 
			// the difference between any existing tenant in the reader
			// catalog will be deleted (i.e. this are descriptors that exist
			// in the reader tenant, but not in the from tenant which we are 
			// replicating).
___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747168903:-O6RC2mHA0myQs44fIdU:b9au9an) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L66)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

nit: i think the logic would be easier to follow if `desc` were renamed `fromDesc`, given that you also call `ForEachDescriptor` on `allExistingDescs` below
</blockquote></details>

Good point, let me change this.

Done.
___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 83 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747197761:-O6RCOMYDSIa8gzI0lRB:b29xue5) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L83)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

^ actually, my check above :
```
existingDesc.GetParentID()  == desc.GetParentID()
```
could lead to false positives if the system database id is not hard coded at start up
</blockquote></details>

Done.

Good point. System database ID is fixed and hard coded. So, we can just confirm that it matches.
___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747150226:-O6REWZG06qgClyD_ruL:b9ddulp) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L100)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

it's kinda sad that we have to marshal then unmarshall `t.Table` just to place it into the mutBuilder's Table Descriptor. I don't know if i can suggest a better solution, as you're the expert on this stuff. 
</blockquote></details>

Done.

Could just do protoutil.Clone instead, let me do that.
___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747157857:-O6RGSPk5Wr4FOrodMzx:b-pbb4g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L118)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

why don't we set external row data for non physical tables? will reads to those naturally reroute to physical tables which have external row data set?
</blockquote></details>

Yeah, for views the reads will go through the descriptors selected by those tables.
___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747183809:-O6RGt-fAMkPWoAAQUFr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog.go#L188)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

perhaps we default error on an unrecognized type?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/replication/reader_catalog_test.go` line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1747211932:-O6RJd4r1uqLOPqPUmF9:b-1225ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ee9f23402cf25e9b96ecfe4997d6ab16a767c0/pkg/sql/catalog/replication/reader_catalog_test.go#L91)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

could you add a few test cases, where you rerun SetupOrAdvanceStandbyReaderCatalog with a new AOST after:
-  insert into t1
- create a new table and insert data into it
- conduct some schema change on it (e.g. add column in it)
- drop that table new table
</blockquote></details>

I have more tests in the follow on PR...but let me just add these here.

File: pkg/sql/catalog/replication/reader_catalog.go
Comment (msbutler) [inline pkg/sql/catalog/replication/reader_catalog.go:54]: s/this/these/r

File: pkg/sql/catalog/replication/reader_catalog_test.go
Comment (msbutler) [inline pkg/sql/catalog/replication/reader_catalog_test.go:89]: nit: s/systme/system/r

**Review by msbutler — APPROVED** [review]
Looks awesome! I'll let you decide if you want a pair of sql-foundations eyes on this before you merge.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130184)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @azhu-crl, @msbutler, and @stevendanna)

___
*[`pkg/sql/catalog/replication/reader_catalog.go` line 54 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/130184#gh-1752356984:-O6SBTirAz4FwUscqyzZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba833e99b924e3b1c981a1226380199d7927f49e/pkg/sql/catalog/replication/reader_catalog.go#L54)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

s/this/these/r
</blockquote></details>

Done.

**Comment by fqazi** [comment]
@msbutler TFTR!

bors r+

**Pull request merged**
